arangod/RocksDBEngine/RocksDBGeoIndex.cpp

```diff
@@ -263,6 +420,205 @@ class RDBNearIterator final : public IndexIterator {
   std::unique_ptr<rocksdb::Iterator> _iter;
 };
 
+class RDBCoveringIterator final : public IndexIterator {
+ public:
+  /// @brief Construct an RocksDBGeoIndexIterator based on Ast Conditions
+  RDBCoveringIterator(LogicalCollection* collection, transaction::Methods* trx,
+                      RocksDBGeoIndex const* index, geo::QueryParams&& params)
+      : IndexIterator(collection, trx, ReadOwnWrites::no),
+        // geo index never needs to observe own writes since it cannot be used
+        // for an UPSERT subquery
+        _index(index),
+        _covering(std::move(params)) {
+    RocksDBTransactionMethods* mthds =
+        RocksDBTransactionState::toMethods(trx, _collection->id());
+    _iter = mthds->NewIterator(_index->columnFamily(), {});
+    TRI_ASSERT(_index->columnFamily()->GetID() ==
+               RocksDBColumnFamilyManager::get(
+                   RocksDBColumnFamilyManager::Family::GeoIndex)
+                   ->GetID());
+  }
+
+  char const* typeName() const override {
+    return "geo-index-covering-iterator";
+  }
+
+  /// internal retrieval loop
+  template<typename F>
+  inline bool nextToken(F&& cb, size_t limit) {
+    if (_covering.isDone()) {
+      // we already know that no further results will be returned by the index
+      return false;
+    }
+
+    // We keep going, until either we have reached our limit or we have
+    // scanned all intervals delivered by _covering:
+    while (limit > 0 &&
+           (!_covering.isDone() || _scanningInterval < _scan.size())) {
+      if (!_covering.hasNext()) {
+        performScan();
+      }
+
+      while (limit > 0 && _covering.hasNext()) {
+        if (std::forward<F>(cb)(_covering.getNext())) {
+          --limit;
+        }
+        _covering.next();
+      }
+    }
+    return !_covering.isDone() || _scanningInterval < _scan.size();
+  }
+
+  bool nextDocumentImpl(DocumentCallback const& cb, size_t limit) override {
+    return nextToken(
+        [this, &cb](LocalDocumentId const& docid) -> bool {
+          bool result = true;  // this is updated by the callback
+          if (!_collection->getPhysical()
+                   ->read(
+                       _trx, docid,
+                       [&](LocalDocumentId const&, VPackSlice doc) {
+                         geo::FilterType const ft = _covering.filterType();
+                         geo::ShapeContainer const& filter =
+                             _covering.filterShape();
+                         TRI_ASSERT(filter.type() !=
+                                    geo::ShapeContainer::Type::EMPTY);
+                         geo::ShapeContainer test;
+                         Result res = _index->shape(doc, test);
+                         TRI_ASSERT(res.ok());  // this should never fail here
+                         if (res.fail() ||
```

Comment:
could it be removed if the assert should never fail?

Reply:
It could, if we are right and do not have a but which would trigger the assertion. This is just defensive programming: If ever we have a situation in which the `shape` call fails, then in a release build the assertion is ignored. In this case we prefer in production to simply ignore the result rather than handing out a potentially damaged result.